pkg/infrastructure/gcp/clusterapi/iam.go

```diff
@@ -102,48 +105,56 @@ func AddServiceAccountRoles(ctx context.Context, projectID, serviceAccountID str
 		return fmt.Errorf("failed to create resourcemanager service: %w", err)
 	}
 
-	policy, err := getPolicy(ctx, service, projectID)
-	if err != nil {
-		return err
-	}
+	backoff := time.Second
+	for {
+		policy, err := getPolicy(ctx, service, projectID)
+		if isQuotaExceededError(err) {
+			logrus.Debugf("failed to get IAM policy, retrying after backoff")
+			time.Sleep(backoff)
+			continue
+		} else if err != nil {
+			return fmt.Errorf("failed to get IAM policy, unexpected error: %w", err)
+		}
 
-	member := fmt.Sprintf("serviceAccount:%s", serviceAccountID)
-	for _, role := range roles {
-		err = addMemberToRole(policy, role, member)
-		if err != nil {
-			return fmt.Errorf("failed to add role %s to %s: %w", role, member, err)
+		member := fmt.Sprintf("serviceAccount:%s", serviceAccountID)
+		for _, role := range roles {
+			err = addMemberToRole(policy, role, member)
+			if err != nil {
+				return fmt.Errorf("failed to add role %s to %s: %w", role, member, err)
+			}
 		}
-	}
 
-	err = setPolicy(ctx, service, projectID, policy)
-	if err != nil {
-		return err
+		err = setPolicy(ctx, service, projectID, policy)
+		if err == nil {
```

Comment:
nit: maybe restructure so that the default is the nil error case

```suggestion
		if err != nil {
		    if isConflictError(err) {
		        logrus.Debugf("Concurrent IAM policy changes, restarting read/modify/write after %s seconds", backoff)
			time.Sleep(backoff)
			backoff *= 2
			if backoff > 30*time.Second {
				return fmt.Errorf("error applying IAM policy, too many conflicts. Latest error: %w", err)
			}
			continue
		}
		return fmt.Errorf("failed to set IAM policy, unexpected error: %w", err)
	}
	logrus.Debugf("Successfully set IAM policy")
	return nil
```

Reply:
Good point. Changed.